lib/sessions.js

```diff
@@ -219,73 +308,111 @@ class ClientSession extends EventEmitter {
+      return;
+    }
+  } else {
+    if (txnState === TxnState.STARTING_TRANSACTION) {
+      // the transaction was never started, we can safely exit here
+      session.transaction.transition(TxnState.TRANSACTION_ABORTED);
+      callback(null, null);
+      return;
+    }
+
+    if (txnState === TxnState.TRANSACTION_ABORTED) {
+      callback(new MongoError('Cannot call abortTransaction twice'));
+      return;
+    }
+
+    if (
+      txnState === TxnState.TRANSACTION_COMMITTED ||
+      txnState === TxnState.TRANSACTION_COMMITTED_EMPTY
+    ) {
+      callback(new MongoError('Cannot call abortTransaction after calling commitTransaction'));
+      return;
+    }
   }
 
+  // construct and send the command
   const command = { [commandName]: 1 };
-  if (clientSession.transactionOptions.writeConcern) {
-    Object.assign(command, { writeConcern: clientSession.transactionOptions.writeConcern });
-  } else if (clientSession.clientOptions && clientSession.clientOptions.w) {
-    Object.assign(command, { writeConcern: { w: clientSession.clientOptions.w } });
+
+  // apply a writeConcern if specified
+  if (session.transaction.options.writeConcern) {
+    Object.assign(command, { writeConcern: session.transaction.options.writeConcern });
+  } else if (session.clientOptions && session.clientOptions.w) {
+    Object.assign(command, { writeConcern: { w: session.clientOptions.w } });
   }
 
   function commandHandler(e, r) {
-    resetTransactionState(clientSession);
+    if (commandName === 'commitTransaction') {
+      session.transaction.transition(TxnState.TRANSACTION_COMMITTED);
+
+      if (
+        e &&
+        (e instanceof MongoNetworkError ||
+          e instanceof MongoWriteConcernError ||
+          isRetryableError(e))
+      ) {
+        if (e.errorLabels) {
+          const idx = e.errorLabels.indexOf('TransientTransactionError');
+          if (idx !== -1) {
+            e.errorLabels.splice(idx, 1);
+          }
+        } else {
+          e.errorLabels = [];
+        }
+
+        e.errorLabels.push('UnknownTransactionCommitResult');
+      }
+    } else {
+      session.transaction.transition(TxnState.TRANSACTION_ABORTED);
+    }
+
     callback(e, r);
   }
 
+  // The spec indicates that we should ignore all errors on `abortTransaction`
   function transactionError(err) {
     return commandName === 'commitTransaction' ? err : null;
   }
 
   // send the command
-  clientSession.topology.command(
-    'admin.$cmd',
-    command,
-    { session: clientSession },
-    (err, reply) => {
-      if (err && isRetryableError(err)) {
-        return clientSession.topology.command(
-          'admin.$cmd',
-          command,
-          { session: clientSession },
-          (_err, _reply) => commandHandler(transactionError(_err), _reply)
-        );
-      }
-
-      commandHandler(transactionError(err), reply);
+  session.topology.command('admin.$cmd', command, { session }, (err, reply) => {
```

Comment:
This makes me feel weird for some reason, but I can't figure out a good way to refactor this.

I'd love for there to be some way to not write the `session.topology.command` out more than once.

Reply:
I'm going to leave this be, but certainly submit a PR if you come up with something you think is better!